helix-term/src/handlers/document_colors.rs

```diff
@@ -0,0 +1,267 @@
+        .handlers
+        .document_colors
+        .active_requests
+        .insert(view_id, controller);
+
+    tokio::spawn(async move {
+        match cancelable_future(future, cancel).await {
+            Some(Ok(response)) => {
+                job::dispatch(move |editor, _| {
+                    attach_document_colors(editor, view_id, doc_id, response, offset_encoding)
+                })
+                .await;
+            }
+            Some(Err(err)) => log::error!("document color request failed: {err}"),
+            None => (),
+        }
+    });
+}
+
+fn attach_document_colors(
+    editor: &mut Editor,
+    view_id: ViewId,
+    doc_id: DocumentId,
+    response: Option<Vec<lsp::ColorInformation>>,
+    offset_encoding: OffsetEncoding,
+) {
+    if !editor.config().lsp.display_color_swatches || editor.tree.try_get(view_id).is_none() {
+        return;
+    }
+
+    let Some(doc) = editor.documents.get_mut(&doc_id) else {
+        return;
+    };
+
+    let mut doc_colors = match response {
+        Some(colors) if !colors.is_empty() => colors,
+        _ => {
+            doc.set_color_swatches(view_id, DocumentColorSwatches::default());
+            return;
+        }
+    };
+
+    // Most language servers will already send them sorted but ensure this is the case to avoid errors on our end
+    doc_colors.sort_unstable_by_key(|color| color.range.start);
+
+    let mut color_swatches = Vec::with_capacity(doc_colors.len());
+    let mut color_swatches_padding = Vec::with_capacity(doc_colors.len());
+    let mut colors = Vec::with_capacity(doc_colors.len());
+
+    let doc_text = doc.text();
+
+    for ColorInformation { range, color } in doc_colors {
+        let swatch_idx =
+            match helix_lsp::util::lsp_pos_to_pos(doc_text, range.start, offset_encoding) {
+                Some(pos) => pos,
+                // Skip color swatches that have no "real" position
+                None => continue,
+            };
+        color_swatches_padding.push(InlineAnnotation::new(swatch_idx, " "));
+        color_swatches.push(InlineAnnotation::new(swatch_idx, "■"));
+        colors.push(Theme::rgb_highlight(
+            (color.red * 255.) as u8,
+            (color.green * 255.) as u8,
+            (color.blue * 255.) as u8,
+        ));
+    }
+
+    let swatches = DocumentColorSwatches {
+        color_swatches,
+        colors,
+        color_swatches_padding,
+    };
+
+    doc.set_color_swatches(view_id, swatches)
+}
+
+pub(super) fn register_hooks(handlers: &Handlers) {
+    register_hook!(move |event: &mut DocumentDidOpen<'_>| {
+        // when a document is initially opened, request colors for it
+        let views: Vec<_> = event
+            .editor
+            .tree
+            .views()
+            .filter_map(|(view, _)| (view.doc == event.doc).then_some(view.id))
+            .collect();
+
+        for view in views {
+            request_document_colors_for_view(event.editor, view, event.doc);
+        }
```

Comment:
Inlay hints need to track the view and store the hints on a per-view basis because the inlay hint request gives hints just for the requested range. The document colors request covers the whole document though so there's no need to track the view: the colors will apply to any view that shows the document. So the hook doesn't need to track view information and the helper can change from `request_document_colors_for_view(&mut Editor, ViewId, DocumentId)` to `request_document_colors(&mut Editor, DocumentId)`